### ISSUE TITLE
`host.map` explosion

### ISSUE THREAD

**Issue by ocdi** [haproxytech/kubernetes-ingress#265]
I'm using the `1.5.0-rc2` image and I've observed for some reason that the `host.map` grows repeating the same entries many times. There are 8 unique entries in this file but for some reason it duplicates them many times to 4752 entries.

![image](https://user-images.githubusercontent.com/2775804/105018106-aeda4100-5a98-11eb-8d59-59cb8965311a.png)

The trigger for the explosion appears to be adding/removing entries to a service that is also used as an external name service. The service only has one IP address on the endpoints and if that entry is set to not ready (so zero available addresses) and then back to ready, the issue occurs.

This would be _ok_, however it causes all traffic to be routed to the default backend which isn't ideal. Restarting the haproxy controller pod allows it to build a correct `host.map` and resume correct operation.

The 4752 entries only happens after running a while (maybe a few days). After starting a fresh controller and manually adjusting the addresses I can see varying host.map sizes that are smaller and things are working correctly.

```
/etc/haproxy/maps # cat host.map  | wc -l
54
/etc/haproxy/maps # cat host.map  | wc -l
54
/etc/haproxy/maps # cat host.map  | wc -l
54
/etc/haproxy/maps # cat host.map  | wc -l
54
/etc/haproxy/maps # cat host.map  | wc -l
9
/etc/haproxy/maps # cat host.map  | wc -l
9
/etc/haproxy/maps # cat host.map  | wc -l
27
/etc/haproxy/maps # cat host.map  | wc -l
27
/etc/haproxy/maps # cat host.map  | wc -l
27
/etc/haproxy/maps # cat host.map  | wc -l
27
```

I haven't attempted to dig through the code to find a potential cause, which I will try to do in my spare time. This is on a test cluster with test services so isn't a big deal, but it would be good to iron out the cause.

**Comment by Mo3m3n** [comment]
Thanks @ocdi for your report.
This is probably related to how we construct the map files, we append entries without checking for duplicates.
So redundancy is expected (HAProxy will pick the first entry anyway) but I was expecting that the redundancy would be contained and does not grow between sync cycles because we clean the config in between.
I should be able to reproduce this and get back with more information.

> This would be ok, however it causes all traffic to be routed to the default backend which isn't ideal. Restarting the haproxy controller pod allows it to build a correct host.map and resume correct operation.

This part is not clear, having a big map files causes traffic to be routed to default backend ?

**Comment by ocdi** [comment]
The main reason I reported the issue was after the map file got huge it seemed to not work correctly and the traffic only went to the default backend. The only resolution was to restart the pod, in which case everything reset and worked as normal.

**Comment by Mo3m3n** [comment]
So this seems to happen when there is a wrong configuration making HAProxy sync fails, in such case Controller config state is not cleared and leads to infinite loop of Controller trying to re-apply the same erroneous config ( one of the side effects is the continuous increase of the map files) 
I have a potential fix here https://github.com/Mo3m3n/kubernetes-ingress/commit/360045b42b94430a75cac494993d43491f5571ca
Can you give it a try via `haproxytech/kubernetes-ingress:dev` and in any case have a look to the logs to see if there is any sync error.

**Comment by ocdi** [comment]
@Mo3m3n thanks for looking into this. I have tested your fix and that appears to have solved it. 👍 The scenario I described is no longer resulting in the massive map file.

I believe the issue is the externalName has a service with a single endpoint. When that endpoint is not available, the DNS server isn't resolving the given name. That results into this error here being logged:
```
2021-01-20T02:48:55.663268986Z 2021/01/20 02:48:55 ERROR   controller.go:276 14: ERR transactionId=b0f3b12c-2556-49d6-8193-69c69a31ac40 
2021-01-20T02:48:55.663309087Z line=81 msg="'server SRV_1' could not resolve address 'sys-beta.default.svc.cluster.local'."
2021-01-20T02:48:55.663314787Z msg="Failed to initialize server(s) addr."
```

That is consistent with your theory about the error causing a loop.

**Comment by Mo3m3n** [comment]
I see so I have added a patch where controller configures ExtenalName backends with `default-server init-addr last,libc,none` directive this way HAProxy won't failt if DNS is not resolving, you can give it a try by re-pulling`haproxytech/kubernetes-ingress:dev` 
However keep in mind that with Controller version 1.5,  HAProxy won't retry resolving DNS at runtime and you will have to wait for the next reload (this can be triggered by updating a config for example) for HAProxy to do a second lookup.
HAProxy does have a DNS runtime resolver though, we are planning to make use of it starting from Ingress Controller 1.6
If you want to further discuss this don't hesitate to open a dedicated issue.

**Comment by Mo3m3n** [comment]
these fies are available in `haproxytech/kubernetes-ingress:1.5.0`
Thanks again @ocdi  for reporting the map issue, since this is fixed we can discuss the DNS resolve in a dedicated github issue.
